### PR TITLE
[framework] Closed Loop Token

### PR DESCRIPTION
**Pull request by damirka**
## Description 

Adds the `sui::token` module to the framework.

## Test Plan 

To be ported in future updates.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [x] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

- Adds a closed loop `token` to the framework

### CODE REVIEW

**Comment by damirka** [comment]
Updated the PR and removed the owner field (which was intended as TTO prevention).
Thanks @tzakian

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (tzakian) [inline crates/sui-framework/packages/sui-framework/sources/token.move:423]: We may want `request` to be the first argument to make this more amenable to receiver syntax

**Review by tzakian — COMMENTED** [review]
Overall looks good, left a bunch of smaller comments throughout though.

One overarching comment I would make though is when determining the order of arguments in the function is to consider what they will be like with receiver syntax. E.g., in `add_rule`, `rule_config` etc. the `self` argument most likely should be the first argument to provide the most ergonomic usage with receiver syntax.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/token.move:61]: I've given it some thought and decided to wait for he final API to land to have some more time to think about it. The implementation is very straightforward and I've checked that we won't have a conflict there.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/token.move:423]: Does it apply to other functions as well? The issue here is that we always put witness first, and it's already this way in a lot of packages / projects.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (tzakian) [reply crates/sui-framework/packages/sui-framework/sources/token.move:118]: My thinking would be to have something like:

```
enum ActionEnum<T> {
  Spend(Balance<T>),
  Send(address),
  ...
} 
```
Basically a variant for each action that you want to support in here so all the different options aren't needed (but nothing better we can do for now than what you have right now).

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (tzakian) [reply crates/sui-framework/packages/sui-framework/sources/token.move:423]: I would say it's worth taking an eye to all these functions with receiver syntax in-mind (although these were the ones that popped out the most to me). 

Might be worth bucking the trend of the existing ones since before we didn't really have that good of a reason whereas now there's more of a reason to change the order of these arguments possibly? 

I don't have super strong opinions here though and whether the order should be changed since there's definitely an argument to be made for consistency as well.

File: crates/sui-framework/packages/sui-framework/tests/pay_tests.move
Comment (damirka) [inline crates/sui-framework/packages/sui-framework/tests/pay_tests.move:1]: This is just a minor patch in `sui::pay` tests to get rid of `public entry`.

**Review by amnn — COMMENTED** [review]
Overall looks great!  Main thing for me (although I think you're already working on these) is tests demonstrating the full flow for the examples.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/token.move:118]: Noice. If it could be upgradable then yes, enums...

File: examples/move/token/sources/gems.move
Comment (damirka) [reply examples/move/token/sources/gems.move:84]: Yeah, it does.

File: examples/move/token/sources/regulated_coin.move
Comment (damirka) [reply examples/move/token/sources/regulated_coin.move:249]: That's the feature I desperately want in dummy context. We can't set sender today.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/token.move:96]: I've no opinion, ascii::String feels a bit old too me and I don't believe we use it much to not cause confusion.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/token.move:443]: I just added it everywhere because there's been moments of regret for not being able to add it in the future.

File: examples/move/token/sources/regulated_coin.move
Comment (damirka) [reply examples/move/token/sources/regulated_coin.move:252]: Do you want it to be identical?

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (amnn) [reply crates/sui-framework/packages/sui-framework/sources/token.move:443]: But there are some public functions that don't have it, that's why I asked what your heuristic was.

File: examples/move/token/sources/regulated_coin.move
Comment (amnn) [reply examples/move/token/sources/regulated_coin.move:252]: Identical would be fine because it's prefixed by the package name.  Exactly opposite would be fine as well, e.g. `denylist::EUserDenied`, `allowlist::EUserNotAllowed`, but the two similar terms that weren't the same was confusing to me.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/token.move:443]: Two come to mind - `mint` and `burn` which are limited to admin and we don't expect advanced functionality there which could require sender / context checks.

File: examples/move/token/sources/coffee.move
Comment (damirka) [inline examples/move/token/sources/coffee.move:1]: You can easily imagine this being integrated into a small real Coffee Shop flow where a backend process would listen to "CoffeePurchased" event and allow showing the QR code or some proof of transaction while buying a coffee. 

Something like WeChatPay and similar QR + App combos widely used in different regions.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (manolisliolios) [inline crates/sui-framework/packages/sui-framework/sources/token.move:198]: I think we can replace `t` with `token` for better naming?

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (manolisliolios) [inline crates/sui-framework/packages/sui-framework/sources/token.move:445]: Since we're using this assertion multiple times for access control, should we move it to a function with a more descriptive name?

Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/token.move]: Then you'd either have a function `assert_cap_matches(policy, cap)` which is not great as this function is not very reusable. Or redo an assertion everywhere as `assert!(cap_matches(policy, cap), ENotAuthorized)`. Both feel like an overkill for something that simple. At least for now.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/token.move:201]: Why would you do that?

**Review by tzakian — APPROVED** [review]
Looks great to me. Amazing work!

**Comment by tzakian** [comment]
Reverted snapshots for protocol version 32, which are too early to be created since this version hasn't shipped yet (31 is just shipping to testnet). Also reverted a change to the protocol version 31 for deepbook, that doesn't look like it should be there.

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (amnn) [reply crates/sui-framework/packages/sui-framework/sources/token.move:50]: This still seems to be a thing.

**Review by amnn — COMMENTED** [review]
LGTM as well, but, nit: note the error code (not a big issue, of course!)

File: crates/sui-framework/packages/sui-framework/sources/token.move
Comment (tzakian) [reply crates/sui-framework/packages/sui-framework/sources/token.move:50]: Agh, my bad for not seeing this/fixing it before landing this PR. Put up a fix PR here: https://github.com/MystenLabs/sui/pull/14845 cc @damirka

**Pull request merged**
